src/Raven.Server/Documents/Replication/ReplicationLoader.cs

```diff
@@ -235,19 +235,30 @@ private void CreatePullReplicationAsHub(TcpConnectionOptions tcpConnectionOption
                 PullReplicationDefinitionName = initialRequest.PullReplicationDefinitionName
             };
 
+            if (_outgoing.TryAdd(outgoingReplication) == false)
+            {
+                outgoingReplication.Dispose();
+                tcpConnectionOptions.Dispose();
+                return;
+            }
+
             outgoingReplication.Failed += OnOutgoingSendingFailed;
             outgoingReplication.SuccessfulTwoWaysCommunication += OnOutgoingSendingSucceeded;
             outgoingReplication.SuccessfulReplication += ResetReplicationFailuresInfo;
-            _outgoing.TryAdd(outgoingReplication); // can't fail, this is a brand new instance
 
             outgoingReplication.StartPullReplicationAsHub(tcpConnectionOptions.Stream, supportedVersions);
             OutgoingReplicationAdded?.Invoke(outgoingReplication);
         }
 
-        public void RunPullReplicationAsSink(TcpConnectionOptions tcpConnectionOptions, JsonOperationContext.ManagedPinnedBuffer buffer, PullReplicationAsSink destination)
+        public void RunPullReplicationAsSink(
+            TcpConnectionOptions tcpConnectionOptions, 
+            JsonOperationContext.ManagedPinnedBuffer buffer, 
+            PullReplicationAsSink destination, 
+            OutgoingReplicationHandler source)
         {
             var newIncoming = CreateIncomingReplicationHandler(tcpConnectionOptions, buffer, destination.HubDefinitionName);
             newIncoming.Failed += RetryPullReplication;
+            _outgoing.TryRemove(source);
```

Comment:
can you explain why?

Reply:
We are pulling and therefore incoming, upon failure 'RetryPullReplication' will put us back as an outgoing